clang/lib/Sema/SemaTemplate.cpp

```diff
@@ -2258,6 +2258,94 @@ class ExtractTypeForDeductionGuide
   }
 };
 
+// Build a deduction guide with the specified parameter types.
+FunctionTemplateDecl *
+buildDeductionGuide(Sema &SemaRef, TemplateDecl *OriginalTemplate,
+                    TemplateParameterList *TemplateParams,
+                    CXXConstructorDecl *Ctor, ExplicitSpecifier ES,
+                    TypeSourceInfo *TInfo, SourceLocation LocStart,
+                    SourceLocation Loc, SourceLocation LocEnd, bool IsImplicit,
+                    llvm::ArrayRef<TypedefNameDecl *> MaterializedTypedefs = {}) {
+  DeclContext *DC = OriginalTemplate->getDeclContext();
+  auto DeductionGuideName =
+      SemaRef.Context.DeclarationNames.getCXXDeductionGuideName(
+          OriginalTemplate);
+
+  DeclarationNameInfo Name(DeductionGuideName, Loc);
+  ArrayRef<ParmVarDecl *> Params =
+      TInfo->getTypeLoc().castAs<FunctionProtoTypeLoc>().getParams();
+
+  // Build the implicit deduction guide template.
+  auto *Guide =
+      CXXDeductionGuideDecl::Create(SemaRef.Context, DC, LocStart, ES, Name,
+                                    TInfo->getType(), TInfo, LocEnd, Ctor);
+  Guide->setImplicit(IsImplicit);
+  Guide->setParams(Params);
+
+  for (auto *Param : Params)
+    Param->setDeclContext(Guide);
+  for (auto *TD : MaterializedTypedefs)
+    TD->setDeclContext(Guide);
+
+  auto *GuideTemplate = FunctionTemplateDecl::Create(
+      SemaRef.Context, DC, Loc, DeductionGuideName, TemplateParams, Guide);
+  GuideTemplate->setImplicit(IsImplicit);
+  Guide->setDescribedFunctionTemplate(GuideTemplate);
+
+  if (isa<CXXRecordDecl>(DC)) {
+    Guide->setAccess(AS_public);
+    GuideTemplate->setAccess(AS_public);
+  }
+
+  DC->addDecl(GuideTemplate);
+  return GuideTemplate;
+}
+
+// Transform a given template type parameter `TTP`.
+TemplateTypeParmDecl *
+transformTemplateTypeParam(Sema &SemaRef, DeclContext *DC,
+                           TemplateTypeParmDecl *TTP,
+                           MultiLevelTemplateArgumentList &Args,
+                           unsigned NewDepth, unsigned NewIndex) {
+  // TemplateTypeParmDecl's index cannot be changed after creation, so
+  // substitute it directly.
+  auto *NewTTP = TemplateTypeParmDecl::Create(
+      SemaRef.Context, DC, TTP->getBeginLoc(), TTP->getLocation(), NewDepth,
+      NewIndex, TTP->getIdentifier(), TTP->wasDeclaredWithTypename(),
+      TTP->isParameterPack(), TTP->hasTypeConstraint(),
+      TTP->isExpandedParameterPack()
+          ? std::optional<unsigned>(TTP->getNumExpansionParameters())
+          : std::nullopt);
+  if (const auto *TC = TTP->getTypeConstraint())
+    SemaRef.SubstTypeConstraint(NewTTP, TC, Args,
+                                /*EvaluateConstraint*/ true);
```

Comment:
```suggestion
                                /*EvaluateConstraint=*/ true);
```
Do we mean to evaluate the constraint?  Typically we don't want to evaluate the constraint unless we are checking the decl for a call, or for constraint comparisons.  So we really don't ever substitute these ever (and the substituted versions don't really end up in the AST).

Reply:
Yes, I think so, though the function parameter name is a bit confusing.

Here, we intend to *rewrite* the `TTP->getTypeConstraint()` (update the old template occurrences with the new template parameters). Looking at the `SubstTypeConstraint` implementation, if we set it to false, it will [attach](https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/SemaTemplateInstantiate.cpp#L2886-L2890) the old constraint to `NewTTP`.

(This is a refactoring change which lifts the existing function from `ConvertConstructorToDeductionGuideTransform` for code sharing purpose).

Reply:
> For the most part, we WANT the 'old' constraint to attach. All the work to instantiate the constraints end up going against the original template.

I think this is not we want for this case, we're in the process of building a new `FunctionTemplate` from the original template here, specifically building the `TemplateTypeParmDecl` for the new template. Attaching the `old` constraint to it is not correct because this will make the instantiation of the new template go against the template parameter in original template, and will lead to crashes in the normal CTAD cases, e.g. 

```
template <typename A, typename T>
concept True = true;
struct concepts {
  template <class Y>
  struct B {
    template <True<Y> Z> B(Y, Z);
  };
};

concepts::B b(0, 1);
```

Reply:
Hmm... but as-written it seems like it would cause problems where the original function-template isn't instantiate-able.  It seems to me that instead of substituting it, perhaps we should be 'converting' it in some way to update the references?  I'll have to think on it, though I'm @zygoloid  can come around and answer for us.

Reply:
This case is special, we're performing a template substitution with dependent template args. I think it is fine, the constraint satisfaction check is [not performed](https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/SemaTemplate.cpp#L5029) for dependent args.

Reply:
Hmm... if it is always dependent template args we're putting in that part should be fine.  At that point it becomes just a 'replace args' instead of a substitution, which should be ok.

However, are we making sure we're replacing them 1:1 (including parent templates)?  Else the depth of some of them is going to 'change' and later substitution is going to fail.

Reply:
Yeah, we're replacing them 1:1.